projects/ngx-sub-form/src/lib/ngx-sub-form.component.ts

```diff
@@ -66,6 +74,63 @@ export abstract class NgxSubFormComponent<ControlInterface, FormInterface = Cont
 
   private controlKeys: (keyof FormInterface)[] = [];
 
+  // instead of having the validators defined in the utils file
+  // we define them here to have access to the form types
+
+  // `ngxSubFormValidators` should be used at a group level validation
+  // with the `getFormGroupControlOptions` hook
+  protected ngxSubFormValidators = {
```

Comment:
While I understand this pattern is following the angular pattern of `namespace.validatorname`, I think that concept by angular is actually a bad idea, because it is unfriendly to tree shakers - if you use one validator, youre basically guaranteed to have to import them all. I think a simple pure function per validator, in it's own file is the most tree shaking friendly option

Reply:
I started exactly how you described it.
But then realized we'd have to pass the types as argument =/
Not impossible... not really awful but slightly more heavy :man_shrugging:?

Although, I do have to admit that it feels wrong to have validators defined within the class and require the class to be instantiated before being able to use them.

Reply:
Making another pass on this one, do we really care about tree shaking here? I wouldn't expect ngx-sub-form to embed plenty of validators (if so they should really be extracted into a separated package IMO). Plus, I'm not sure that using a namespace is better for tree shaking :thinking: but yes could have functions where we need to pass the type to.